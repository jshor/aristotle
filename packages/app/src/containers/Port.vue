<template>
  <port-pivot :rotation="rotation">
    <div v-if="showHelper">
      <port-helper orientation="horizontal" />
      <port-helper orientation="vertical" />
    </div>

    <draggable
      v-if="draggable"
      :position="draggablePosition"
      :data-id="id"
      :class="{ snappable }"
      snap=".port"
      @drag="onDrag"
      @dragStart="dragStart"
      @dragEnd="dragEnd"
      class="port">
      <port-handle
        :active="snappable"
        @mousedown="mousedown"
        @mouseup="mouseup"
      />
    </draggable>
    <port-handle
      v-else
      :active="snappable"
      @mousedown="mousedown"
      @mouseup="mouseup"
    />
  </port-pivot>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { mapActions, mapGetters, mapState } from 'vuex'
import IPoint from '../interfaces/IPoint'
import Draggable from './Draggable.vue'
import PortHandle from '../components/PortHandle.vue'
import PortHelper from '../components/PortHelper.vue'
import PortPivot from '../components/PortPivot.vue'
import getAncestor from '../utils/getAncestor'

export default defineComponent({
  components: {
    Draggable,
    PortHandle,
    PortHelper,
    PortPivot
  },
  props: {
    /**
     * Port ID.
     */
    id: String,

    /**
     * List of sibling ports (all ports on the parent element).
     */
    siblings: {
      type: Object,
      default: () => ({})
    },

    /**
     * Unit circle rotation value (0, 1, 2, 3).
     */
    rotation: {
      type: Number,
      default: 0
    },

    /**
     * Position of the port on the canvas.
     */
    position: {
      type: Object as PropType<IPoint>,
      default: () => ({
        x: 0,
        y: 0
      })
    },

    /**
     * Whether or not the port is draggable.
     */
    draggable: {
      type: Boolean,
      default: false
    },

    /**
     * Whether or not the snap helpers should be enabled.
     * Should be set to true when the port on the opposite end of a connection is being dragged.
     */
    showHelper: {
      type: Boolean,
      default: false
    },

    /**
     * The port type.
     * - 0 = output
     * - 1 = input
     * - 2 = freeport
     */
    type: {
      type: Number,
      default: 0
    },

    /**
     * Port directional orientation.
     */
    orientation: {
      type: Number,
      default: 0
    }
  },
  data () {
    return {
      /**
       * The temporary ID of the draggable port.
       */
      cloneId: 'DRAGGED_PORT',

      /**
       * The relative dragged position of the dragged port.
       */
      draggablePosition: {
        x: 0,
        y: 0
      } as IPoint,

      /**
       * The absolute position of the dragged port.
       */
      absolutePosition: {
        x: 0,
        y: 0
      } as IPoint
    }
  },
  computed: {
    ...mapGetters([
      'zoom'
    ]),
    ...mapState([
      'activePort'
    ]),

    /**
     * Whether or not this port can accept an incoming dragged port.
     * This will update the UI to make the port bigger when true.
     */
    snappable () {
      return this.activePort !== null // && activePort.type !== this.type
    }
  },
  mounted () {
    window.addEventListener('mouseup', this.mouseup)
  },
  destroy () {
    window.removeEventListener('mouseup', this.mouseup)
  },
  methods: {
    ...mapActions([
      'updatePortPositions',
      'setActivePort',
      'connect',
      'disconnect',
      'showPortSnapHelpers',
      'hidePortSnapHelpers'
    ]),

    /**
     * Sets the absolute position of the static (i.e., not dragged) port.
     */
    setAbsolutePosition () {
      const { x, y } = this.$el.getBoundingClientRect()
      const editor = getAncestor(this, 'canvas')

      this.absolutePosition = editor.fromDocumentToEditorCoordinates({ x, y })
    },

    /**
     * Tells the store to show the snap helpers of ports adjacent to this one.
     */
    mousedown () {
      if (this.type === 2) {
        const siblingIds = Object
          .values(this.siblings)
          .reduce((ids: string[], s: any) => [
            ...ids,
            ...s.map(({ id }) => id)
          ], [])

        this.showPortSnapHelpers(siblingIds)
      }
    },

    /**
     * Tells the store that the port's no longer active, and to hide the helpers.
     */
    mouseup () {
      this.setActivePort(null)
      this.hidePortSnapHelpers()
    },

    /**
     * Initializes the draggable connection.
     */
    dragStart ({ position }) {
      this.setAbsolutePosition()
      this.setActivePort({
        id: this.cloneId,
        position: {
          x: 0,
          y: 0
        },
        type: this.type,
        orientation: this.orientation
      })

      this.draggablePosition = {
        x: NaN,
        y: NaN
      }
      this.onDrag(position)

      this.connect({
        source: this.id,
        target: this.cloneId
      })
    },

    /**
     * Updates the store position with the absolute position of the port.
     */
    onDrag ({ position }) {
      if (!position) return
      this.updatePortPositions({
        [this.cloneId]: {
          position: {
            // port coordinates are stored absolutely (relative to its existing position)
            x: position.x + this.absolutePosition.x,
            y: position.y + this.absolutePosition.y
          },
          id: this.cloneId,
          type: this.type,
          orientation: this.orientation,
          rotation: 0
        }
      })
    },

    /**
     * Handles terminating the temporary active port, and connects the port to the one at its dragged location.
     * This method handles all results of a user-driven port-dragging interaction.
     */
    dragEnd ({ position, snappedId }) {
      this.draggablePosition = {
        x: 0,
        y: 0
      }
      this.disconnect({
        source: this.id,
        target: this.cloneId
      })
      this.connect({
        source: this.id,
        target: snappedId
      })
      this.setActivePort(null)
    }
  }
})
</script>
