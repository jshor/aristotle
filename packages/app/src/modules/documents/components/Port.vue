<template>
  <div class="port" :title="rotation" :style="{
    transform: `rotate(-${rotation * 90}deg)`
  }">
    <!-- <div class="port__helper port__helper--vertical snappable" />
    <div class="port__helper port__helper--horizontal snappable" /> -->
    <div class="port__handle"  />
      <draggable class="port" @drag="onDrag" @dragStart="dragStart" @dragEnd="dragEnd" :position="draggablePosition" :data-id="id" :class="{ snappable }" v-if="draggable">
        <div class="port__handle" @mousedown="mousedown" @mouseup="mouseup" :class="{ 'port__handle--active': snappable }"  />
      </draggable>
      <slot />
    <!-- </div> -->
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator'
import Draggable from './Draggable.vue'
import { Getter } from '../store/decorators'

@Component({
  components: {
    Draggable
  }
})
export default class Port extends Vue {

  @Getter('documents', 'zoom')
  public zoom: number

  @Prop()
  public id: string

  @Prop({ default: 0 })
  public rotation: any

  @Prop({ default: () => ({
    x: 0, y: 0
  }) })
  public position: any

  @Prop({ default: true })
  public draggable: boolean

  @Prop({ default: 0 })
  public type: number

  @Prop({ default: 0 })
  public orientation: number

  public cloneId: string = 'DRAGGED_PORT'

  public draggablePosition: any = {
    x: 0,
    y: 0
  }

  public absolutePosition: any = {
    x: 0,
    y: 0
  }

  get snappable () {
    const { activePort } = this.$store.state.documents

    return activePort !== null // && activePort.type !== this.type
  }

  setAbsolutePosition () {
    const parent = (this.$parent.$parent.$parent.$parent.$refs.canvas as any).getBoundingClientRect()
    const { x, y } = this.$el.getBoundingClientRect()

    this.absolutePosition = {
      x: x - (parent.x * this.zoom),
      y: y - (parent.y * this.zoom)
    }
  }

  mousedown () {
    this.setAbsolutePosition()
    this.setActivePort({
      id: this.cloneId,
      position: {
        x: 0,
        y: 0
      },
      type: this.type,
      orientation: this.orientation
    })
  }

  mouseup () {
    this.$store.dispatch('setActivePort', null)
  }

  dragStart ({ position }) {
    this.draggablePosition = {}
    this.onDrag(position)

    this.$store.dispatch('connect', {
      source: this.id,
      target: this.cloneId
    })
  }

  onDrag ({ position }) {
    if (!position) return
    this.$store.dispatch('updatePortPositions', {
      [this.cloneId]: {
        position: {
          x: position.x + this.absolutePosition.x,
          y: position.y + this.absolutePosition.y
        },
        id: this.cloneId,
        type: this.type,
        orientation: this.orientation,
        rotation: 0
      }
    })
  }

  dragEnd ({ position, snappedId }) {
    this.draggablePosition = {
      x: 0,
      y: 0
    }

    this.$store.dispatch('disconnect', {
      source: this.id,
      target: this.cloneId
    })
    this.$store.dispatch('connect', {
      source: this.id,
      target: snappedId
    })
    this.$store.dispatch('setActivePort', null)
  }

  setActivePort (port) {
    this.$store.dispatch('setActivePort', port)
  }
}
</script>

<style lang="scss">
.dr {
  background:blue;
}
.port {
  width: 0;
  height: 0;
  position: relative;

  &__helper {
    content: ' ';
    display: block;
    position: absolute;
    // background-color: rgba(255,0,0,1);

    &--vertical {
      width: 1px;
      left: -1px;
      top: -100vh;
      height: 200vh;
    }

    &--horizontal {
      height: 1px;
      left: -100vh;
      width: 200vh;
    }
  }


  &--rotated {
    transform: rotate(-90deg);
  }

  &__handle {
    position: absolute;
    display: flex;
    justify-content: center;
    align-items: center;
    border: 1px solid black;
    box-sizing: border-box;
    top: -8px;
    left: -8px;
    width: 16px;
    height: 16px;
    background-color: red;
    border-radius: 50%;
    cursor: grab;
    transition: all 0.25s;
    z-index: 1001;

    &:active {
      cursor: grabbing;
    }

    &--active {
      width: 24px;
      height: 24px;
      top: -12px;
      left: -12px;
    }
  }
}
</style>
