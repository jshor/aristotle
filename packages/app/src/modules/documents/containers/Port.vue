<template>
  <port-pivot :rotation="rotation">
    <div v-if="showHelper">
      <port-helper orientation="horizontal" />
      <port-helper orientation="vertical" />
    </div>

    <draggable
      v-if="draggable"
      :position="draggablePosition"
      :data-id="id"
      :class="{ snappable }"
      snap=".port"
      @drag="onDrag"
      @dragStart="dragStart"
      @dragEnd="dragEnd"
      class="port">
      <port-handle
        :active="snappable"
        @mousedown="mousedown"
        @mouseup="mouseup"
      />
    </draggable>
    <port-handle
      v-else
      :active="snappable"
      @mousedown="mousedown"
      @mouseup="mouseup"
    />
  </port-pivot>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator'
import Draggable from './Draggable.vue'
import PortHandle from '../components/PortHandle.vue'
import PortHelper from '../components/PortHelper.vue'
import PortPivot from '../components/PortPivot.vue'
import { Getter } from '../store/decorators'
import IPoint from '../../../interfaces/IPoint'
import getAncestor from '../../../utils/getAncestor'
import Canvas from './Documents.vue'

@Component({
  components: {
    Draggable,
    PortHandle,
    PortHelper,
    PortPivot
  }
})
export default class Port extends Vue {
  /**
   * Current zoom value of the editor.
   *
   * @type {number}
   */
  @Getter('documents', 'zoom')
  public zoom: number

  /**
   * Port ID.
   *
   * @type {string}
   */
  @Prop()
  public id: string

  /**
   * List of sibling ports (all ports on the parent element).
   *
   * @type {IPort[]}
   */
  @Prop({ default: () => ({}) })
  public siblings: any[]

  /**
   * Unit circle rotation value (0, 1, 2, 3).
   *
   * @type {number}
   */
  @Prop({ default: 0 })
  public rotation: any

  /**
   * Position of the port on the canvas.
   *
   * @type {IPoint}
   */
  @Prop({
    default: () => ({
      x: 0,
      y: 0
    })
  })
  public position: IPoint

  /**
   * Whether or not the port is draggable.
   *
   * @type {boolean}
   */
  @Prop({ default: true })
  public draggable: boolean

  /**
   * Whether or not the snap helpers should be enabled.
   * Should be set to true when the port on the opposite end of a connection is being dragged.
   *
   * @type {boolean}
   */
  @Prop({ default: false })
  public showHelper: boolean

  /**
   * The port type.
   * - 0 = output
   * - 1 = input
   * - 2 = freeport
   *
   * @type {number}
   */
  @Prop({ default: 0 })
  public type: number

  /**
   * Port directional orientation.
   *
   * @type {number}
   */
  @Prop({ default: 0 })
  public orientation: number

  /**
   * The temporary ID of the draggable port.
   *
   * @type {string}
   */
  public cloneId: string = 'DRAGGED_PORT'

  /**
   * The relative dragged position of the dragged port.
   *
   * @type {IPoint}
   */
  public draggablePosition: IPoint = {
    x: 0,
    y: 0
  }

  /**
   * The absolute position of the dragged port.
   *
   * @type {IPoint}
   */
  public absolutePosition: IPoint = {
    x: 0,
    y: 0
  }

  /**
   * Whether or not this port can accept an incoming dragged port.
   * This will update the UI to make the port bigger when true.
   *
   * @type {boolean}
   */
  get snappable () {
    const { activePort } = this.$store.state.documents

    return activePort !== null // && activePort.type !== this.type
  }

  mounted () {
    window.addEventListener('mouseup', this.mouseup)
  }

  destroy () {
    window.removeEventListener('mouseup', this.mouseup)
  }

  /**
   * Sets the absolute position of the static (i.e., not dragged) port.
   */
  setAbsolutePosition () {
    const { x, y } = this.$el.getBoundingClientRect()
    const editor = getAncestor(this, Canvas) as Canvas

    this.absolutePosition = editor.fromDocumentToEditorCoordinates({ x, y })
  }

  /**
   * Tells the store to show the snap helpers of ports adjacent to this one.
   */
  mousedown () {
    if (this.type === 2) {
      const siblingIds = Object
        .values(this.siblings)
        .reduce((ids, s) => [
          ...ids,
          ...s.map(({ id }) => id)
        ], [])

      this.$store.dispatch('showPortSnapHelpers', siblingIds)
    }
  }

  /**
   * Tells the store that the port's no longer active, and to hide the helpers.
   */
  mouseup () {
    this.$store.dispatch('setActivePort', null)
    this.$store.dispatch('hidePortSnapHelpers')
  }

  /**
   * Initializes the draggable connection.
   *
   * @param {}
   */
  dragStart ({ position }) {
    this.setAbsolutePosition()
    this.setActivePort({
      id: this.cloneId,
      position: {
        x: 0,
        y: 0
      },
      type: this.type,
      orientation: this.orientation
    })

    this.draggablePosition = {
      x: NaN,
      y: NaN
    }
    this.onDrag(position)

    this.$store.dispatch('connect', {
      source: this.id,
      target: this.cloneId
    })
  }

  /**
   * Updates the store position with the absolute position of the port.
   *
   * @param {*}
   */
  onDrag ({ position }) {
    if (!position) return
    this.$store.dispatch('updatePortPositions', {
      [this.cloneId]: {
        position: {
          // port coordinates are stored absolutely (relative to its existing position)
          x: position.x + this.absolutePosition.x,
          y: position.y + this.absolutePosition.y
        },
        id: this.cloneId,
        type: this.type,
        orientation: this.orientation,
        rotation: 0
      }
    })
  }

  /**
   * Handles terminating the temporary active port, and connects the port to the one at its dragged location.
   * This method handles all results of a user-driven port-dragging interaction.
   *
   * @param {*}
   */
  dragEnd ({ position, snappedId }) {
    this.draggablePosition = {
      x: 0,
      y: 0
    }
    this.$store.dispatch('disconnect', {
      source: this.id,
      target: this.cloneId
    })
    this.$store.dispatch('connect', {
      source: this.id,
      target: snappedId
    })
    this.$store.dispatch('setActivePort', null)
  }

  /**
   * Tells the store which port is being actively-dragged.
   *
   * @param {Port} port
   */
  setActivePort (port) {
    this.$store.dispatch('setActivePort', port)
  }
}
</script>
