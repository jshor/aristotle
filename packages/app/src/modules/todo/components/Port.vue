<template>
  <div class="port" :title="rotation" :style="{
    transform: `rotate(-${rotation * 90}deg)`
  }">
    <div class="port__handle" :class="{
    'port__handle--active': active
  }">
      <draggable class="port snappable" @drag="onDrag" @dragStart="dragStart" @dragEnd="dragEnd" :position="draggablePosition" :data-id="id" snap="port--input">
        <div class="port__handle dr" />
      </draggable>
      <slot />
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator'
import Draggable from './Draggable.vue'

@Component({
  components: {
    Draggable
  }
})
export default class Port extends Vue {

  @Prop()
  public id!: string

  @Prop({ default: 0 })
  public rotation: any

  @Prop({ default: false })
  public active: boolean

  public cloneId: string = 'DRAGGED_PORT'

  public draggablePosition: any = {
    x: 0,
    y: 0
  }

  dragStart ({ position }) {
    const { x, y } = this.$el.getBoundingClientRect()

    this.draggablePosition = {}
    this.onDrag(position)

    this.$emit('dragStart', {
      source: this.id,
      target: this.cloneId
    })
  }

  onDrag ({ position }) {
    this.$emit('drag', {
      [this.cloneId]: {
        position,
        id: this.cloneId
      }
    })
  }

  dragEnd ({ position, snappedId }) {
    this.draggablePosition = {
      x: 0,
      y: 0
    }

    this.$emit('dragEnd', {
      source: this.id,
      oldTarget: this.cloneId,
      newTarget: snappedId
    })
  }
}
</script>

<style lang="scss">
.port {
  width: 0;
  height: 0;
  position: relative;

  &--rotated {
    transform: rotate(-90deg);
  }

  &__handle {
    position: absolute;
    display: flex;
    justify-content: center;
    align-items: center;
    top: -8px;
    left: -8px;
    width: 16px;
    height: 16px;
    background-color: red;
    border-radius: 50%;
    cursor: pointer;
    transition: all 0.25s;
  z-index: 1001;

    &--active {
      width: 32px;
      height: 32px;
      top: -16px;
      left: -16px;
    }
  }
}
</style>
