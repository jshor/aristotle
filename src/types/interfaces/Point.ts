interface Point {
  /**
   * X-axis coordinate.
   *
   * @type {number}
   */
  x: number

  /**
   * Y-axis coordinate.
   *
   * @type {number}
   */
  y: number
}

export default Point
