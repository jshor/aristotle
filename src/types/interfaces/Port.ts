import BinaryWavePulse from '@/store/document/oscillator/BinaryWavePulse'
import PortType from '../enums/PortType'
import Direction from '../enums/Direction'
import Point from './Point'

interface Port {
  id: string
  name: string
  elementId: string
  virtualElementId?: string
  position: Point
  type: PortType
  rotation: number
  orientation: Direction
  value: number
  hue: number
  isFreeport: boolean
  isMonitored: boolean
  connectedPortIds: string[]
  wave?: BinaryWavePulse
}

export default Port
