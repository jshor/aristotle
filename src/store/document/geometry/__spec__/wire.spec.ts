import Point from '@/types/interfaces/Point'
import {
  inflectDirection,
  getPortDirection,
  getEndpoints,
  default as renderLayout
} from '../wire'
import Port from '@/types/interfaces/Port'

describe('Wire Layout', () => {
  const createPort = (position: Point): Port => ({
    id: '',
    name: '',
    connectedPortIds: [],
    elementId: '',
    value: 0,
    isFreeport: false,
    type: 1,
    orientation: 0,
    rotation: 0,
    position,
    hue: 0,
    isMonitored: false
  })

  describe('renderLayout()', () => {
    it('should render a top-left (a) to bottom-right wire (b)', () => {
      const a: Point = { x: 20, y: 65 }
      const b: Point = { x: 30, y: 30 }

      expect(renderLayout(createPort(a), createPort(b))).toMatchSnapshot()
    })

    it('should render a bottom-left (a) to top-right wire (b)', () => {
      const a: Point = { x: 50, y: 30 }
      const b: Point = { x: 20, y: 65 }

      expect(renderLayout(createPort(a), createPort(b))).toMatchSnapshot()
    })

    it('should render a bottom-left (b) to top-right wire (a)', () => {
      const a: Point = { x: 50, y: 30 }
      const b: Point = { x: 20, y: 65 }

      expect(renderLayout(createPort(a), createPort(b))).toMatchSnapshot()
    })

    it('should render a top-left (b) to bottom-right wire (a)', () => {
      const a: Point = { x: 50, y: 65 }
      const b: Point = { x: 30, y: 30 }

      expect(renderLayout(createPort(a), createPort(b))).toMatchSnapshot()
    })
  })

  describe('inflectDirection()', () => {
    describe('when the y-value of `a` exceeds the y-value of `b`', () => {
      const a: Point = { x: 0, y: 10 }
      const b: Point = { x: 0, y: 5 }

      it('should return 2 when the direction is 0', () => {
        expect(inflectDirection(0, a, b)).toEqual(2)
      })

      it('should return 0 when the direction is 2', () => {
        expect(inflectDirection(2, a, b)).toEqual(0)
      })
    })

    describe('when the y-value of `a` is less than the y-value of `b`', () => {
      const a: Point = { x: 0, y: 5 }
      const b: Point = { x: 0, y: 10 }

      it('should return 0 when the direction is 0', () => {
        expect(inflectDirection(0, a, b)).toEqual(0)
      })

      it('should return 2 when the direction is 2', () => {
        expect(inflectDirection(2, a, b)).toEqual(2)
      })
    })

    describe('when the x-value of `a` is less than the x-value of `b`', () => {
      const a: Point = { x: 10, y: 0 }
      const b: Point = { x: 15, y: 0 }

      it('should return 3 when the direction is 1', () => {
        expect(inflectDirection(1, a, b)).toEqual(3)
      })

      it('should return 1 when the direction is 3', () => {
        expect(inflectDirection(3, a, b)).toEqual(1)
      })
    })

    describe('when the x-value of `a` exceeds the x-value of `b`', () => {
      const a: Point = { x: 15, y: 0 }
      const b: Point = { x: 10, y: 0 }

      it('should return 1 when the direction is 1', () => {
        expect(inflectDirection(1, a, b)).toEqual(1)
      })

      it('should return 3 when the direction is 3', () => {
        expect(inflectDirection(3, a, b)).toEqual(3)
      })
    })

    describe('when the direction is outside the interval [0,3]', () => {
      it('should return its transformed, rotated equivalent in the range [0,3]', () => {
        const a: Point = { x: 10, y: 0 }
        const b: Point = { x: 15, y: 0 }

        expect(inflectDirection(1, a, b)).toEqual(3)
      })

      it('should return its rotated equivalent in the range [0,3] when no transformations apply', () => {
        const a: Point = { x: 15, y: 0 }
        const b: Point = { x: 10, y: 0 }

        expect(inflectDirection(5, a, b)).toEqual(1)
      })
    })
  })

  describe('getPortDirection()', () => {
    const basePort: Port = createPort({ x: 0, y: 0 })

    it('should set the direction to be 1 less than the addition of the port\'s orientation and rotation', () => {
      const a: Point = { x: 10, y: 0 }
      const b: Point = { x: 15, y: 0 }

      const port = {
        ...basePort,
        type: 1,
        orientation: 1,
        rotation: 2
      }

      expect(getPortDirection(port, a, b)).toEqual(2)
    })

    describe('when the port is a FreePort', () => {
      describe('when the rotation is rotated to one of the top two quadrants', () => {
        const a: Point = { x: 10, y: 0 }
        const b: Point = { x: 15, y: 0 }

        it('should return 0 when the rotation is in the first quadrant', () => {
          const port = {
            ...basePort,
            type: 2,
            orientation: 1,
            rotation: 2
          }

          expect(getPortDirection(port, a, b)).toEqual(2)
        })

        it('should return 0 when the rotation is in the second quadrant', () => {
          const port = {
            ...basePort,
            type: 2,
            orientation: 1,
            rotation: 0
          }

          expect(getPortDirection(port, a, b)).toEqual(0)
        })
      })

      describe('when the rotation is rotated to the third or fourth quadrants', () => {
        const a: Point = { x: 10, y: 0 }
        const b: Point = { x: 15, y: 0 }

        it('should return 0 when the rotation is in the third quadrant', () => {
          const port = {
            ...basePort,
            type: 2,
            orientation: 2,
            rotation: 1
          }

          expect(getPortDirection(port, a, b)).toEqual(2)
        })

        it('should return 2 when the rotation is in the fourth quadrant', () => {
          const port = {
            ...basePort,
            type: 2,
            orientation: 3,
            rotation: 2
          }

          expect(getPortDirection(port, a, b)).toEqual(0)
        })
      })
    })
  })

  describe('getEndpoints()', () => {
    describe('when the x-value of `a` is less than or equal to the x-value of `b`', () => {
      describe('when the y-value of `a` is less than or equal to the y-value of `b`', () => {
        const a: Point = { x: 20, y: 30 }
        const b: Point = { x: 50, y: 65 }

        it('should return (0, 0) for the start point', () => {
          expect(getEndpoints(a, b).start).toEqual({ x: 0, y: 0 })
        })

        it('should return (30, 35) for the start point', () => {
          expect(getEndpoints(a, b).end).toEqual({ x: 30, y: 35 })
        })
      })

      describe('when the y-value of `a` exceeds the y-value of `b`', () => {
        const a: Point = { x: 20, y: 65 }
        const b: Point = { x: 30, y: 30 }

        it('should return (0, 35) for the start point', () => {
          expect(getEndpoints(a, b).start).toEqual({ x: 0, y: 35 })
        })

        it('should return (10, 0) for the start point', () => {
          expect(getEndpoints(a, b).end).toEqual({ x: 10, y: 0 })
        })
      })
    })

    describe('when the x-value of `a` exceeds the x-value of `b`', () => {
      describe('when the y-value of `a` is less than or equal to the y-value of `b`', () => {
        const a: Point = { x: 50, y: 30 }
        const b: Point = { x: 20, y: 65 }

        it('should return (30, 0) for the start point', () => {
          expect(getEndpoints(a, b).start).toEqual({ x: 30, y: 0 })
        })

        it('should return (0, 35) for the start point', () => {
          expect(getEndpoints(a, b).end).toEqual({ x: 0, y: 35 })
        })
      })

      describe('when the y-value of `a` exceeds the y-value of `b`', () => {
        const a: Point = { x: 50, y: 65 }
        const b: Point = { x: 30, y: 30 }

        it('should return (20, 35) for the start point', () => {
          expect(getEndpoints(a, b).start).toEqual({ x: 20, y: 35 })
        })

        it('should return (0, 0) for the start point', () => {
          expect(getEndpoints(a, b).end).toEqual({ x: 0, y: 0 })
        })
      })
    })
  })
})
