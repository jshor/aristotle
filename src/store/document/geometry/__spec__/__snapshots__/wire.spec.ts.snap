// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Wire Layout > renderLayout() > should render a bottom-left (a) to top-right wire (b) 1`] = `
{
  "height": 65,
  "minX": -22.5,
  "minY": -7.5,
  "path": "M 30 0 C 15 0 -15 35 0 35",
  "width": 75,
}
`;

exports[`Wire Layout > renderLayout() > should render a bottom-left (b) to top-right wire (a) 1`] = `
{
  "height": 65,
  "minX": -22.5,
  "minY": -7.5,
  "path": "M 30 0 C 15 0 -15 35 0 35",
  "width": 75,
}
`;

exports[`Wire Layout > renderLayout() > should render a top-left (a) to bottom-right wire (b) 1`] = `
{
  "height": 65,
  "minX": -17.5,
  "minY": -7.5,
  "path": "M 0 35 C -10 35 0 0 10 0",
  "width": 50,
}
`;

exports[`Wire Layout > renderLayout() > should render a top-left (b) to bottom-right wire (a) 1`] = `
{
  "height": 65,
  "minX": -17.5,
  "minY": -7.5,
  "path": "M 20 35 C 10 35 -10 0 0 0",
  "width": 60,
}
`;

exports[`Wire Layout renderLayout() should render a bottom-left (a) to top-right wire (b) 1`] = `
{
  "height": 65,
  "minX": -22.5,
  "minY": -7.5,
  "path": "M 30 0 C 15 0 -15 35 0 35",
  "width": 75,
}
`;

exports[`Wire Layout renderLayout() should render a bottom-left (b) to top-right wire (a) 1`] = `
{
  "height": 65,
  "minX": -22.5,
  "minY": -7.5,
  "path": "M 30 0 C 15 0 -15 35 0 35",
  "width": 75,
}
`;

exports[`Wire Layout renderLayout() should render a top-left (a) to bottom-right wire (b) 1`] = `
{
  "height": 65,
  "minX": -17.5,
  "minY": -7.5,
  "path": "M 0 35 C -10 35 0 0 10 0",
  "width": 50,
}
`;

exports[`Wire Layout renderLayout() should render a top-left (b) to bottom-right wire (a) 1`] = `
{
  "height": 65,
  "minX": -17.5,
  "minY": -7.5,
  "path": "M 20 35 C 10 35 -10 0 0 0",
  "width": 60,
}
`;
